pkg/vetter/conflictingvirtualservicehost/vet.go

```diff
@@ -48,63 +79,323 @@ type hostAndGateway struct {
+		for _, route := range vs.Spec.GetHttp() {
+			for _, match := range route.GetMatch() {
+				addRouteToMergedVsTree(trie, match.GetUri(), vs)
+			}
+		}
+	}
+	return trie
+}
+
+// Add a particular route to the route trie. If the given route already has a route rule,
+// add it to the list of route rules for the given node/route.
+func addRouteToMergedVsTree(trie *routeTrie, match *istiov1alpha3.StringMatch, vs *v1alpha3.VirtualService) {
+	current := trie
+	rRule := getRouteRuleFromMatch(match, vs)
+
+	// Regexs are treated as exceptions to the trie construction rule.
+	// This is largely due to the complexities in determining whether two regexs
+	// conflict.
+	if rRule.ruleType == regex {
+		trie.regexs = append(trie.regexs, rRule)
+		return
+	}
+
+	if strings.HasSuffix("/", rRule.route) {
+		rRule.route = strings.TrimSuffix(rRule.route, "/")
+	}
+
+	// Routes have leading slashes, remove the leading empty string from the array after the split
+	components := strings.Split(rRule.route, "/")[1:]
+	for count, component := range components {
+		if next, ok := current.subRoutes[component]; ok {
+			if count == len(components)-1 {
+				next.routeRules = append(next.routeRules, rRule)
+			} else {
+				current = next
+			}
+		} else {
+			newSubRoutes := make(map[string]*routeTrie)
+			// This is the final component in a route rule and a new node is created for it
+			if count == len(components)-1 {
+				current.subRoutes[component] = &routeTrie{subRoutes: newSubRoutes, routeRules: []routeRule{rRule}}
+			} else {
+				newSubRoute := &routeTrie{subRoutes: newSubRoutes, routeRules: []routeRule{}}
+				current.subRoutes[component] = newSubRoute
+				current = newSubRoute
+			}
+		}
+	}
+}
+
+func conflictingSubroutes(trie *routeTrie, rRule routeRule, conflictingRules [][]routeRule) ([][]routeRule, error) {
+	for _, rule := range trie.routeRules {
+		if c, err := conflict(rRule, rule); err != nil {
+			return conflictingRules, err
+		} else {
+			if c {
+				conflictingRules = append(conflictingRules, []routeRule{rRule, rule})
+			}
+		}
+	}
+
+	for _, descendant := range trie.subRoutes {
+		if len(descendant.routeRules) == 0 {
+			if c, err := conflictingSubroutes(descendant, rRule, conflictingRules); err != nil {
+				return conflictingRules, err
+			} else {
+				conflictingRules = append(conflictingRules, c...)
+			}
+		} else {
+			// Recurse down but carefully! We want to report all conflicts and
+			// we'll skip potential conflicts with the current route rule if we
+			// recurse in the previous for loop (with the descendant rule as the "rRule" variable),
+			for _, rule := range append(trie.routeRules, rRule) {
+				newRules, err := conflictingSubroutes(descendant, rule, conflictingRules)
+				if err != nil {
+					return conflictingRules, err
+				}
+				conflictingRules = newRules
+			}
+		}
+	}
+	return conflictingRules, nil
+}
+func addConflictsForSameRoute(trie *routeTrie, conflictingRules [][]routeRule) [][]routeRule {
```

Comment:
Are we sure we always want rules for the same route to conflict? This is not a real conflict if in the correct order in the same VS